nav2-usability/scenario-code/sign-in-routing/router.dart

```diff
@@ -0,0 +1,309 @@
+import 'package:flutter/material.dart';
+
+void main() {
+  runApp(BooksApp());
+}
+
+class Credentials {
+  final String username;
+  final String password;
+
+  Credentials(this.username, this.password);
+}
+
+abstract class Authentication {
+  bool isSignedIn();
+
+  void signOut();
+
+  Future<bool> signIn(String username, String password);
+}
+
+class MockAuthentication implements Authentication {
+  bool _signedIn = false;
+
+  @override
+  bool isSignedIn() {
+    return _signedIn;
+  }
+
+  @override
+  void signOut() {
+    _signedIn = false;
+  }
+
+  @override
+  Future<bool> signIn(String username, String password) async {
+    return _signedIn = true;
+  }
+}
+
+class BooksApp extends StatefulWidget {
+  @override
+  State<StatefulWidget> createState() => _BooksAppState();
+}
+
+class AppState extends ChangeNotifier {
+  final Authentication auth;
+
+  AppState(this.auth);
+
+  Future<bool> signIn(String username, String password) async {
+    var success = await auth.signIn(username, password);
+    notifyListeners();
+    return success;
+  }
+
+  bool _viewingBooks = false;
+
+  bool get viewingBooks => _viewingBooks;
+
+  set viewingBooks(bool value) {
+    _viewingBooks = value;
+    notifyListeners();
+  }
+}
+
+class _BooksAppState extends State<BooksApp> {
+  final BookRouterDelegate _routerDelegate = BookRouterDelegate();
+  final BookRouteInformationParser _routeInformationParser =
+  BookRouteInformationParser();
+
+  @override
+  void dispose() {
+    _routerDelegate.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp.router(
+      title: 'Books App',
+      routerDelegate: _routerDelegate,
+      routeInformationParser: _routeInformationParser,
+    );
+  }
+}
+
+class BookRouteInformationParser extends RouteInformationParser<AppRoutePath> {
+  @override
+  Future<AppRoutePath> parseRouteInformation(
```

Comment:
This method should check whether users isSignedIn and return SignInRoutePath if they are not

Reply:
Ah, gotcha.